mcp/src/test/java/io/modelcontextprotocol/server/McpAsyncServerExchangeTests.java

```diff
@@ -0,0 +1,697 @@
+
+		StepVerifier.create(exchange.listRoots()).assertNext(result -> {
+			// Verify the accumulated result is correct
+			assertThat(result.roots()).hasSize(2);
+
+			// Verify that the returned list is unmodifiable
+			assertThatThrownBy(() -> result.roots().add(new McpSchema.Root("file:///test", "Test")))
+				.isInstanceOf(UnsupportedOperationException.class);
+
+			// Verify that clear() also throws UnsupportedOperationException
+			assertThatThrownBy(() -> result.roots().clear()).isInstanceOf(UnsupportedOperationException.class);
+
+			// Verify that remove() also throws UnsupportedOperationException
+			assertThatThrownBy(() -> result.roots().remove(0)).isInstanceOf(UnsupportedOperationException.class);
+		}).verifyComplete();
+	}
+
+	@Test
+	void testGetClientCapabilities() {
+		assertThat(exchange.getClientCapabilities()).isEqualTo(clientCapabilities);
+	}
+
+	@Test
+	void testGetClientInfo() {
+		assertThat(exchange.getClientInfo()).isEqualTo(clientInfo);
+	}
+
+	// ---------------------------------------
+	// Logging Notification Tests
+	// ---------------------------------------
+
+	@Test
+	void testLoggingNotificationWithNullMessage() {
+		StepVerifier.create(exchange.loggingNotification(null)).verifyErrorSatisfies(error -> {
+			assertThat(error).isInstanceOf(McpError.class).hasMessage("Logging message must not be null");
+		});
+	}
+
+	@Test
+	void testLoggingNotificationWithAllowedLevel() {
+
+		McpSchema.LoggingMessageNotification notification = McpSchema.LoggingMessageNotification.builder()
+			.level(McpSchema.LoggingLevel.ERROR)
+			.logger("test-logger")
+			.data("Test error message")
+			.build();
+
+		when(mockSession.sendNotification(eq(McpSchema.METHOD_NOTIFICATION_MESSAGE), eq(notification)))
+			.thenReturn(Mono.empty());
+
+		StepVerifier.create(exchange.loggingNotification(notification)).verifyComplete();
+	}
+
+	@Test
+	void testLoggingNotificationWithFilteredLevel() {
+		// Given - Set minimum level to WARNING, send DEBUG message
+		exchange.setMinLoggingLevel(McpSchema.LoggingLevel.WARNING);
+
+		McpSchema.LoggingMessageNotification debugNotification = McpSchema.LoggingMessageNotification.builder()
+			.level(McpSchema.LoggingLevel.DEBUG)
+			.logger("test-logger")
+			.data("Debug message that should be filtered")
+			.build();
+
+		// When & Then - Should complete without sending notification
+		StepVerifier.create(exchange.loggingNotification(debugNotification)).verifyComplete();
+	}
+
+	@Test
+	void testLoggingNotificationLevelFiltering() {
+		// Given - Set minimum level to WARNING
+		exchange.setMinLoggingLevel(McpSchema.LoggingLevel.WARNING);
+
+		// Test DEBUG (should be filtered)
+		McpSchema.LoggingMessageNotification debugNotification = McpSchema.LoggingMessageNotification.builder()
+			.level(McpSchema.LoggingLevel.DEBUG)
+			.logger("test-logger")
+			.data("Debug message")
+			.build();
+
+		StepVerifier.create(exchange.loggingNotification(debugNotification)).verifyComplete();
```

Comment:
Would it make sense to investigate the mock that it had no interactions for this? Making the mock a spy?

Reply:
addressed